contracts/roles.js

```diff
@@ -0,0 +1,917 @@
+    && !api.assert(typeof roleId === 'string', 'invalid roleId')) {
+    return;
+  }
+  const role = await api.db.findOne('roles', { _id: api.BigNumber(roleId).toNumber() });
+  if (!api.assert(role, 'role does not exist')) return;
+  const inst = await api.db.findOne('instances', { _id: role.instanceId });
+
+  let authorizedCreation = true;
+  if (inst.candidateFee) {
+    const feeTokenBalance = await api.db.findOneInTable('tokens', 'balances', { account: api.sender, symbol: inst.candidateFee.symbol });
+    authorizedCreation = api.BigNumber(inst.candidateFee.amount).lte(0) || api.sender === api.owner
+      ? true
+      : feeTokenBalance && api.BigNumber(feeTokenBalance.balance).gte(inst.candidateFee.amount);
+  }
+
+  const existingApply = await api.db.findOne('candidates', { roleId: role._id, account: api.sender });
+  if (api.assert(authorizedCreation, 'you must have enough tokens to cover the application fee')
+    && api.assert(!existingApply, 'sender already applied for role')) {
+    const newCandidate = {
+      roleId: role._id,
+      account: api.sender,
+      active: true,
+      approvalWeight: { $numberDecimal: '0' },
+    };
+    const insertedId = await api.db.insert('candidates', newCandidate);
+
+    if (api.sender !== api.owner && inst.candidateFee) {
+      if (inst.candidateFee.method === 'burn') {
+        await api.executeSmartContract('tokens', 'transfer', {
+          to: 'null', symbol: inst.candidateFee.symbol, quantity: inst.candidateFee.amount,
+        });
+      } else if (inst.candidateFee.method === 'issuer') {
+        const feeTokenObj = await api.db.findOneInTable('tokens', 'tokens', { symbol: inst.candidateFee.symbol });
+        await api.executeSmartContract('tokens', 'transfer', {
+          to: feeTokenObj.issuer, symbol: inst.candidateFee.symbol, quantity: inst.candidateFee.amount,
+        });
+      }
+    }
+
+    api.emit('applyForRole', { roleId: role._id, candidateId: insertedId._id });
+  }
+};
+
+actions.setApplyActive = async (payload) => {
+  const {
+    roleId,
+    active,
+    isSignedWithActiveKey,
+  } = payload;
+
+  if (!api.assert(isSignedWithActiveKey === true, 'you must use a transaction signed with your active key')
+    && !api.assert(typeof roleId === 'string', 'invalid roleId')) {
+    return;
+  }
+  const role = await api.db.findOne('roles', { _id: api.BigNumber(roleId).toNumber() });
+  const existingApply = await api.db.findOne('candidates', { roleId: role._id, account: api.sender });
+  if (api.assert(role, 'role does not exist')
+    && api.assert(existingApply, 'candidate does not exist for sender')) {
+    existingApply.active = !!active;
+    await api.db.update('candidates', existingApply);
+    api.emit('setApplyActive', { roleId: role._id, account: existingApply.account, active });
+  }
+};
+
+// deposit
+
+async function updateTokenBalances(role, token, quantity) {
+  const upRole = role;
```

Comment:
Why use this `upRole` variable here instead of just directly using the passed in `role` ?